ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/tasks/CompleteBlocksTaskTest.java

```diff
@@ -79,6 +95,115 @@ public void shouldCompleteWithoutPeersWhenAllBlocksAreEmpty() {
+    final ProtocolSpec mockShanghaiSpec = Mockito.mock(ProtocolSpec.class);
+    final WithdrawalsProcessor mockWithdrawalsProcessor = Mockito.mock(WithdrawalsProcessor.class);
+
+    final BlockHeader header1 =
+        new BlockHeaderTestFixture().number(1).withdrawalsRoot(null).buildHeader();
+    final BlockHeader header2 =
+        new BlockHeaderTestFixture().number(2).withdrawalsRoot(Hash.EMPTY_TRIE_HASH).buildHeader();
+
+    when(mockProtocolSchedule.getByBlockHeader((eq(header1)))).thenReturn(mockParisSpec);
+    when(mockParisSpec.getWithdrawalsProcessor()).thenReturn(Optional.empty());
+    when(mockProtocolSchedule.getByBlockHeader((eq(header2)))).thenReturn(mockShanghaiSpec);
+    when(mockShanghaiSpec.getWithdrawalsProcessor())
+        .thenReturn(Optional.of(mockWithdrawalsProcessor));
+
+    final Block block1 =
+        new Block(header1, new BlockBody(Collections.emptyList(), Collections.emptyList()));
+    final Block block2 =
+        new Block(
+            header2,
+            new BlockBody(
+                Collections.emptyList(),
+                Collections.emptyList(),
+                Optional.of(Collections.emptyList())));
+
+    final List<Block> expectedBlocks = asList(block1, block2);
+    final EthTask<List<Block>> task =
+        CompleteBlocksTask.forHeaders(
+            mockProtocolSchedule,
+            ethContext,
+            List.of(header1, header2),
+            maxRetries,
+            new NoOpMetricsSystem());
+    assertThat(task.run()).isCompletedWithValue(expectedBlocks);
+  }
+
+  @Test
+  public void shouldCompleteWhenBlocksAreEmptyExceptForWithdrawals() {
+    final ProtocolSchedule mockProtocolSchedule = Mockito.mock(ProtocolSchedule.class);
+    final ProtocolSpec mockParisSpec = Mockito.mock(ProtocolSpec.class);
+    final ProtocolSpec mockShanghaiSpec = Mockito.mock(ProtocolSpec.class);
+    final WithdrawalsProcessor mockWithdrawalsProcessor = Mockito.mock(WithdrawalsProcessor.class);
+
+    final Withdrawal withdrawal =
+        new Withdrawal(UInt64.ONE, UInt64.ONE, Address.fromHexString("0x1"), GWei.ONE);
+    final List<Withdrawal> withdrawals = List.of(withdrawal);
+    final Hash withdrawalsRoot = BodyValidation.withdrawalsRoot(withdrawals);
+
+    final BlockHeader preWithdrawalsHeader = new BlockHeaderTestFixture().number(1).buildHeader();
+    final BlockHeader firstWithdrawalsHeader =
+        new BlockHeaderTestFixture().number(2).withdrawalsRoot(withdrawalsRoot).buildHeader();
+    final BlockHeader secondWithdrawalsHeader =
+        new BlockHeaderTestFixture().number(3).withdrawalsRoot(Hash.EMPTY_TRIE_HASH).buildHeader();
+
+    final Block block1 = new Block(preWithdrawalsHeader, BlockBody.empty());
+    final Block block2 =
+        new Block(
+            firstWithdrawalsHeader,
+            new BlockBody(
+                Collections.emptyList(), Collections.emptyList(), Optional.of(withdrawals)));
+    final Block block3 =
```

Comment:
nit: rename to empyWithdrawalsBlock?

Reply:
arguably some value in keeping header1, block1, blockNumber = 1 all in sync